.github/workflows/python-tests.yml

```
@@ -0,0 +1,72 @@
+name: Python Tests
+
+# 1. Run on every push 
+on:
+  push:
+    branches:
+      - '**'
+  pull_request:
+    branches:
+      - main
+
+jobs:
+  test:
+    # 2. Use the latest Ubuntu runner
+    runs-on: ubuntu-latest
+
+    # 3. Spin up a PostgreSQL service for tests
+    services:
+      postgres:
+        image: postgres:15
+        env:
+          POSTGRES_USER: postgres
+          POSTGRES_PASSWORD: postgres
+          POSTGRES_DB: test_db
+        ports:
+          - 5432:5432
+        options: >-
+          --health-cmd "pg_isready -U postgres"
+          --health-interval 10s
+          --health-timeout 5s
+          --health-retries 5
+
+    steps:
+      # 4. Check out code
+      - uses: actions/checkout@v3
+
+      # 5. Wait for Postgres to be ready
+      - name: Wait for Postgres
+        run: |
+          until pg_isready -h localhost -p 5432 -U postgres; do
+            echo "Waiting for Postgres…"
+            sleep 1
+          done
+
+      # 6. Create test database & apply schema
+      - name: Initialize database
+        env:
+          PGPASSWORD: postgres
+        run: |
+          psql -h localhost -U postgres -c "DROP DATABASE IF EXISTS test_db;"
+          psql -h localhost -U postgres -c "CREATE DATABASE test_db;"
+          psql -h localhost -U postgres -d test_db -f sql/01_create-tables.sql
+          psql -h localhost -U postgres -d test_db -f sql/03_create-view.sql
+
+      # 7. Set up Python
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.9'
+
+      # 8. Install dependencies
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install pytest psycopg2-binary python-dotenv
+
+      # 9. Run pytest
+      - name: Run tests
+        env:
+          DATABASE_URL: postgresql://postgres:postgres@localhost:5432/test_db
+        run: |
+          pytest --maxfail=1 --disable-warnings -v
```

`--maxfail=1` can be helpful in making local testing faster, but CI is a pain to re-run, so it's helpful to get as much information as possible when there's a failure


```suggestion
          pytest --disable-warnings -v
```

---

warnings often indicate future failure eg if a feature is being deprecated. `--disable-warnings` should probably be removed and the relevant warnings fixed later